src/runtime-support.cc

```diff
@@ -231,6 +235,74 @@ void rt::reject_protocol(const std::string& reason) {
         throw ValueUnavailable("no current connection available");
 }
 
+void rt::protocol_begin(const std::optional<std::string>& analyzer) {
+    auto cookie = static_cast<Cookie*>(hilti::rt::context::cookie());
+    assert(cookie);
+
+    auto c = std::get_if<cookie::ProtocolAnalyzer>(cookie);
+    if ( ! c )
+        throw ValueUnavailable("no current connection available");
+
+    if ( analyzer ) {
+        auto child = ::zeek::analyzer_mgr->InstantiateAnalyzer(analyzer->c_str(), c->analyzer->Conn());
+        if ( ! child )
+            throw ZeekError(::hilti::rt::fmt("unknown analyzer '%s' requested", *analyzer));
+
+        if ( ! c->analyzer->AddChildAnalyzer(child) )
+            // AddChildAnalyzer() will have deleted child
+            throw ZeekError(::hilti::rt::fmt("could not add analyzer '%s' to connection", *analyzer));
+    }
+    else {
+        // Use a Zeek PIA stream analyzer performing DPD.
+        auto child = new ::zeek::analyzer::pia::PIA_TCP(c->analyzer->Conn());
+
+        if ( ! c->analyzer->AddChildAnalyzer(child) )
+            // AddChildAnalyzer() will have deleted child
+            throw ZeekError(::hilti::rt::fmt("could not add DPD analyzer to connection"));
```

Comment:
This leaks `child`. Let's either clean it up here, it instead use some smart pointer to hold it.

Reply:
See the comment. :-)

Reply:
🤦